llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp

```diff
@@ -3651,6 +3651,11 @@ bool AMDGPUDAGToDAGISel::SelectVOP3PMadMixModsImpl(SDValue In, SDValue &Src,
       // TODO: Should we try to look for neg/abs here?
     }
 
+    // Prevent unnecessary subreg COPY to VGPR_16
+    if (Src.getOpcode() == ISD::TRUNCATE &&
+        Src.getOperand(0).getValueType().getSizeInBits() == 32) {
```

Comment:
Can you try to break this, and find a test where the source type isn't 32?

Reply:
Hi Matt, for fake16 mode and pre-gfx11 target, there are cases that trunc is from f64->f32. 

If we guard this condition with useRealTrue16Insts, then we don't need to check for sizeinbit here

Reply:
Hi Matt. It seems there are cases that the trunc is from i64 to i16.  i.e.  frem with v4f16 case in frem.ll. and thus this check is needed